File: src/main/java/artisra/topinterview150/hashmap/GroupAnagrams.java
package artisra.topinterview150.hashmap;

import org.jetbrains.annotations.NotNull;

import java.util.*;

/**
 * A class for grouping anagrams in a given array of strings.
 *
 * Anagrams are words that contain the same characters, but in a different order.
 * This class efficiently groups anagrams together using a HashMap.
 * @see <a href="https://leetcode.com/problems/group-anagrams/description/">Group Anagrams on LeetCode</a>
 */
public class GroupAnagrams {

    /**
     * Groups anagrams in the given array of strings.
     *
     * @param strings The array of strings to group.
     * @return A list of lists, where each inner list contains anagrams.
     */
    public List<List<String>> groupAnagrams(@NotNull String[] strings) {
        Map<String, List<String>> anagramGroups = new HashMap<>();

        for (String word : strings) {
            String sortedWord = sortString(word);
            if (anagramGroups.containsKey(sortedWord)) {
                anagramGroups.get(sortedWord).add(word);
            } else {
                anagramGroups.put(sortedWord, new ArrayList<>(){{add(word);}});
            }
        }

        return anagramGroups.values().stream().toList();
    }

    /**
     * Sorts the characters of a given string alphabetically.
     *
     * @param string The string to sort.
     * @return The sorted string.
     */
    private String sortString(String string) {
        char[] sortedChars = string.toCharArray();
        Arrays.sort(sortedChars);
        return new String(sortedChars);
    }
}

File: src/main/kotlin/topinterview150/hashmap/GroupAnagrams.kt
package dev.artisra.topinterview150.hashmap

/**
 * A class for grouping anagrams in a given array of strings.
 *
 * Anagrams are words that contain the same characters, but in different orders.
 * This class efficiently groups anagrams together using a hash map.
 * @see <a href="https://leetcode.com/problems/group-anagrams/description/">Group Anagrams on LeetCode</a>
 */
class GroupAnagrams {

    /**
     * Groups anagrams in the given array of strings.
     *
     * @param words the array of strings to group
     * @return a list of lists, where each inner list contains anagrams
     */
    fun groupAnagrams(words: Array<String>): List<List<String>> {
        val anagramGroups = buildMap<String, MutableList<String>> {
            for (word in words) {
                val sortedWord = word.toSorted()
                val anagramList = this.getOrDefault(sortedWord, mutableListOf())
                anagramList.add(word)
                this[sortedWord] = anagramList
            }
        }

        return anagramGroups.values.toList()
    }

    companion object {

        /**
         * Sorts the characters of a string alphabetically.
         *
         * This helper function is used to create a unique identifier for each anagram group.
         *
         * @param this the string to sort
         * @return the sorted string
         */
        private fun String.toSorted(): String {
            val sortedChars = this.toCharArray()
            sortedChars.sort()
            return sortedChars.joinToString("")
        }
    }
}
